File: backend/core/src/modules/document-versions/document-versions.service.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { ProjectAccessService } from '../../common/access/project-access.service';
import { Document, DocumentVersion, Project, WritingGoal } from '../../entities';
import { SearchService } from '../search/search.service';

@Injectable()
export class DocumentVersionsService {
  constructor(
    @InjectRepository(DocumentVersion)
    private readonly versionsRepo: Repository<DocumentVersion>,
    @InjectRepository(Document)
    private readonly documentsRepo: Repository<Document>,
    @InjectRepository(Project)
    private readonly projectsRepo: Repository<Project>,
    @InjectRepository(WritingGoal)
    private readonly goalsRepo: Repository<WritingGoal>,
    private readonly searchService: SearchService,
    private readonly projectAccessService: ProjectAccessService,
  ) {}

  private computeWordCount(content?: string | null): number {
    if (!content) return 0;
    const plain = content
      .replace(/<[^>]*>/g, ' ')
      .replace(/&nbsp;/g, ' ')
      .replace(/\s+/g, ' ')
      .trim();

    if (!plain) return 0;
    const tokens = plain.match(/[\p{L}\p{N}]+/gu);
    return tokens ? tokens.length : 0;
  }

  private async refreshProjectWordCount(projectId: string): Promise<number> {
    const words = await this.documentsRepo
      .createQueryBuilder('d')
      .select('COALESCE(SUM(d.word_count), 0)', 'sum')
      .where('d.project_id = :projectId', { projectId })
      .getRawOne();

    const wordCount = Number(words?.sum ?? 0);
    await Promise.all([
      this.projectsRepo.update({ id: projectId } as any, { wordCount } as any),
      this.goalsRepo.update({ projectId } as any, { currentWords: wordCount } as any),
    ]);
    return wordCount;
  }

  private async assertDocumentAccess(userId: string, documentId: string): Promise<Document> {
    const doc = await this.documentsRepo.findOne({ where: { id: documentId } });
    if (!doc) {
      throw new BadRequestException('문서를 찾을 수 없습니다.');
    }
    await this.projectAccessService.assertProjectAccess(userId, doc.projectId);
    return doc;
  }

  async findAllForUser(userId: string, documentId: string) {
    await this.assertDocumentAccess(userId, documentId);
    return this.versionsRepo.find({
      where: { documentId } as any,
      order: { createdAt: 'DESC' } as any,
    });
  }

  async findOneForUser(userId: string, id: string) {
    const version = await this.versionsRepo.findOne({ where: { id } });
    if (!version) return null;
    await this.assertDocumentAccess(userId, version.documentId);
    return version;
  }

  async createForUser(userId: string, dto: Partial<DocumentVersion>) {
    if (!dto.documentId) {
      throw new BadRequestException('documentId가 필요합니다.');
    }
    await this.assertDocumentAccess(userId, dto.documentId);
    const version = this.versionsRepo.create({ ...dto, createdBy: dto.createdBy ?? userId } as any);
    return this.versionsRepo.save(version as any);
  }

  async removeForUser(userId: string, id: string) {
    const version = await this.versionsRepo.findOne({ where: { id } });
    if (!version) return;
    await this.assertDocumentAccess(userId, version.documentId);
    await this.versionsRepo.delete({ id } as any);
  }

  async restoreForUser(userId: string, versionId: string) {
    const version = await this.versionsRepo.findOne({ where: { id: versionId } });
    if (!version) {
      throw new BadRequestException('버전을 찾을 수 없습니다.');
    }

    const doc = await this.assertDocumentAccess(userId, version.documentId);

    await this.versionsRepo.save(
      this.versionsRepo.create({
        documentId: doc.id,
        content: doc.content ?? '',
        wordCount: doc.wordCount,
        versionName: '복원 전 백업',
        createdBy: userId,
      } as any),
    );

    const wordCount = version.wordCount ?? this.computeWordCount(version.content);
    doc.content = version.content;
    doc.wordCount = wordCount;
    const updated = await this.documentsRepo.save(doc as any);

    await Promise.all([
      this.refreshProjectWordCount(updated.projectId),
      this.searchService.indexDocument('documents', updated.id, {
        id: updated.id,
        projectId: updated.projectId,
        title: updated.title,
        content: updated.content,
        type: updated.type,
      }),
    ]);

    return updated;
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/DocumentVersionsService.java
package store.pjcloud.cowrite.core.service;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import store.pjcloud.cowrite.core.entity.Document;
import store.pjcloud.cowrite.core.entity.DocumentVersion;
import store.pjcloud.cowrite.core.repository.DocumentRepository;
import store.pjcloud.cowrite.core.repository.DocumentVersionRepository;
import store.pjcloud.cowrite.core.repository.ProjectRepository;
import store.pjcloud.cowrite.core.repository.WritingGoalRepository;

@Service
public class DocumentVersionsService {
    private final DocumentVersionRepository versionsRepo;
    private final DocumentRepository documentsRepo;
    private final ProjectRepository projectsRepo;
    private final WritingGoalRepository goalsRepo;
    private final SearchService searchService;
    private final ProjectAccessService projectAccessService;

    public DocumentVersionsService(DocumentVersionRepository versionsRepo,
                                   DocumentRepository documentsRepo,
                                   ProjectRepository projectsRepo,
                                   WritingGoalRepository goalsRepo,
                                   SearchService searchService,
                                   ProjectAccessService projectAccessService) {
        this.versionsRepo = versionsRepo;
        this.documentsRepo = documentsRepo;
        this.projectsRepo = projectsRepo;
        this.goalsRepo = goalsRepo;
        this.searchService = searchService;
        this.projectAccessService = projectAccessService;
    }

    private int computeWordCount(String content) {
        if (content == null || content.isBlank()) return 0;
        String plain = content
            .replaceAll("<[^>]*>", " ")
            .replace("&nbsp;", " ")
            .replaceAll("\\s+", " ")
            .trim();
        if (plain.isBlank()) return 0;
        Pattern pattern = Pattern.compile("[\\p{L}\\p{N}]+", Pattern.UNICODE_CHARACTER_CLASS);
        Matcher matcher = pattern.matcher(plain);
        int count = 0;
        while (matcher.find()) count++;
        return count;
    }

    private void refreshProjectWordCount(UUID projectId) {
        Long sum = documentsRepo.sumWordCount(projectId);
        int wordCount = sum == null ? 0 : sum.intValue();
        projectsRepo.updateWordCount(projectId, wordCount);
        goalsRepo.updateCurrentWords(projectId, wordCount);
    }

    private Document assertDocumentAccess(UUID userId, UUID documentId) {
        Document doc = documentsRepo.findById(documentId).orElse(null);
        if (doc == null) {
            throw new IllegalArgumentException("문서를 찾을 수 없습니다.");
        }
        projectAccessService.assertProjectAccess(userId, doc.getProjectId());
        return doc;
    }

    public List<DocumentVersion> findAllForUser(UUID userId, UUID documentId) {
        assertDocumentAccess(userId, documentId);
        return versionsRepo.findByDocumentIdOrderByCreatedAtDesc(documentId);
    }

    public DocumentVersion findOneForUser(UUID userId, UUID id) {
        DocumentVersion version = versionsRepo.findById(id).orElse(null);
        if (version == null) return null;
        assertDocumentAccess(userId, version.getDocumentId());
        return version;
    }

    @Transactional
    public DocumentVersion createForUser(UUID userId, DocumentVersion dto) {
        if (dto.getDocumentId() == null) {
            throw new IllegalArgumentException("documentId가 필요합니다.");
        }
        assertDocumentAccess(userId, dto.getDocumentId());
        if (dto.getCreatedBy() == null) dto.setCreatedBy(userId);
        return versionsRepo.save(dto);
    }

    @Transactional
    public void removeForUser(UUID userId, UUID id) {
        DocumentVersion version = versionsRepo.findById(id).orElse(null);
        if (version == null) return;
        assertDocumentAccess(userId, version.getDocumentId());
        versionsRepo.deleteById(id);
    }

    @Transactional
    public Document restoreForUser(UUID userId, UUID versionId) {
        DocumentVersion version = versionsRepo.findById(versionId).orElse(null);
        if (version == null) {
            throw new IllegalArgumentException("버전을 찾을 수 없습니다.");
        }

        Document doc = assertDocumentAccess(userId, version.getDocumentId());

        DocumentVersion backup = new DocumentVersion();
        backup.setDocumentId(doc.getId());
        backup.setContent(doc.getContent() == null ? "" : doc.getContent());
        backup.setWordCount(doc.getWordCount());
        backup.setVersionName("복원 전 백업");
        backup.setCreatedBy(userId);
        versionsRepo.save(backup);

        int wordCount = version.getWordCount() != null ? version.getWordCount() : computeWordCount(version.getContent());
        doc.setContent(version.getContent());
        doc.setWordCount(wordCount);
        Document updated = documentsRepo.save(doc);

        refreshProjectWordCount(updated.getProjectId());
        java.util.Map<String, Object> indexed = new java.util.HashMap<>();
        indexed.put("id", updated.getId() == null ? null : updated.getId().toString());
        indexed.put("projectId", updated.getProjectId() == null ? null : updated.getProjectId().toString());
        indexed.put("title", updated.getTitle());
        indexed.put("content", updated.getContent());
        indexed.put("type", updated.getType());
        searchService.indexDocument("documents", updated.getId().toString(), indexed);

        return updated;
    }
}
